### PR TITLE
Consolidate SSL_DEBUG data dump

### PR DESCRIPTION
**Pull request by tmshort**
Clean up SSL_DEBUG data dumping code; use a single, common API.
Send all SSL_DEBUG output to stderr
Use BIO dump routines for consistency

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by mattcaswell** [comment]
I'm wondering whether we should just delete all `SSL_DEBUG` sections instead

**Comment by tmshort** [comment]
I can do that too... I just had a strong desire to clean it up...

**Comment by mattcaswell** [comment]
I'd support their removal but it might be nice to hear what others think.

**Comment by richsalz** [comment]
It's okay to printf in debug mode.  I think the project removed all other library printf's back in 1.1.0  I remember me and Matt worked on that.

**Comment by levitte** [comment]
Interesting...  I've been playing with a small project that adds the possibility to add tracing and debugging hooks, allowing any application to specify callback functions that should get this output instead of printing directly.  We will need *some* mechanism of this sort anyway to allow instrumentation of modules.

**Comment by tmshort** [comment]
I was thinking that there ought to be a debug callbacks, and that this is where these type of print outs should be handled. There already exists a debug callback in the SSL structure (s->ext.debug_cb), which could be extended for this.

**Comment by tmshort** [comment]
(That one, by the way, is set via SSL_CTRL_SET_TLSEXT_DEBUG_CB).

**Comment by levitte** [comment]
I think we'll need something more generic than just for TLS

**Comment by tmshort** [comment]
That one is even more specific for TLSEXT's... but yes

**Comment by richsalz** [comment]
If you add debug hooks, make sure that you can't change the params and affect what is done.

**Comment by levitte** [comment]
> If you add debug hooks, make sure that you can't change the params and affect what is done.

So all input parameters should be `const`.  Yup, that was the intention.

**Comment by tmshort** [comment]
But you can't trust `const`... a user can cast it away. It's debug... I'd memdup the data before passing it to the application callback.

**Comment by levitte** [comment]
Hmmm, what I'm writing won't ever pass data directly to the callback.  It will pass readily formatted text,

**Comment by tmshort** [comment]
Formatted text is not always useful; sometimes developers may want the real data as they debug their applications. They may not want to print but want to examine the data, or have a specific format method (for their own logging).

**Comment by levitte** [comment]
Well...  I don't think we should give them the option of getting actual raw internal data.  We don't today, and I don't think we have a direct future need either.  That's a can of worms I'm not willing to touch.

**Comment by tmshort** [comment]
It would be a copy, not the actual data buffers, as @richsalz suggests. I also think, that for it to be useful, the data would have to be tagged with what it is... and that might also be a can of worms.

**Comment by richsalz** [comment]
I actually wasn't saying make copies of the buffer.  For example "ensure you don't change the data" could be done just via documentation.  And in the first release, _const_ probably suffices.

**Comment by tmshort** [comment]
I guess I'm more paranoid than you...

**Comment by t-j-h** [comment]
It is not our job or responsibility to protect the application programmer or anything that has control of the process from itself. That is outside the scope of our concern. We can document an interface and state we don't expect the parameters to be modified - but there is absolutely no point whatsoever in doing things like taking copies of buffers or trying to enforce things. That just isn't part of our threat model.

File: ssl/ssl_lib.c [hunk 1]
```diff
@@ -5582,3 +5582,15 @@ void SSL_set_allow_early_data_cb(SSL *s,
     s->allow_early_data_cb = cb;
     s->allow_early_data_cb_data = arg;
 }
+
+#ifdef SSL_DEBUG
+/* 
+ * Debug routine, uses the prefix with a possible int argument |ival|
+ */
+void ssl_debug_data_dump(const char* prefix, unsigned int ival, void *data, size_t len)
```
Comment (paulidale) [inline ssl/ssl_lib.c]: Suggest making this: `void ssl_debug_data_dump(const char *prefix, void *data, size_t len, ...)`
Add `__attribute__ ((format (printf, 1, 4)))` on platforms where this is available.  Essentially, don't force a single integer argument which might or might not be used.

Comment (tmshort) [reply ssl/ssl_lib.c]: Since this was wrapped with a macro, I was avoiding using varargs. But I suppose it depends on the CPP.

**Comment by tmshort** [comment]
I updated it with some va_arg stuff....
However, I still see (at least) 3 options to resolve the comments:

1. Keep as is (tidying up and consolidation)
1. Delete all the SSL_DEBUG stuff
1. Add a callback interface

File: ssl/ssl_locl.h [hunk 2]
```diff
@@ -2666,4 +2666,31 @@ void ssl_ctx_system_config(SSL_CTX *ctx);
...
+ */
+# if !defined(__STDC_VERSION__) || __STDC_VERSION__ < 199901L
+#  define SSL_DEBUG_data_dump ssl_debug_data_dump
+# else
+#  ifdef SSL_DEBUG
+#   define SSL_DEBUG_data_dump(...) ssl_debug_data_dump(__VA_ARGS__)
+#  else
+#   define SSL_DEBUG_data_dump(...)
+#  endif
+# endif
```
Comment (levitte) [inline ssl/ssl_locl.h]: Uhmmm, why do we need this?

Comment (tmshort) [reply ssl/ssl_locl.h]: If `SSL_DEBUG` is not defined, the debug code is completely removed. Or, are you referring to something else?

Comment (tmshort) [reply ssl/ssl_locl.h]: (If it's the `__STDC_VERSION__` check, that's based on code from test/testutil.h)

Comment (levitte) [reply ssl/ssl_locl.h]: Ah...  but you're not doing that for C90, so it still looks odd...

Comment (tmshort) [reply ssl/ssl_locl.h]: This code is in test/testutil.h:
```
# if !defined(__STDC_VERSION__) || __STDC_VERSION__ < 199901L
#  define TEST_error         test_error_c90
#  define TEST_info          test_info_c90
# else
#  define TEST_error(...)    test_error(__FILE__, __LINE__, __VA_ARGS__)
#  define TEST_info(...)     test_info(__FILE__, __LINE__, __VA_ARGS__)
# endif
```
So, if THAT code is wrong... then... whoops!

Comment (levitte) [reply ssl/ssl_locl.h]: What happens with the definition of `SSL_DEBUG_data_dump` in C90 when `SSL_DEBUG` is undefined?

**Comment by levitte** [comment]
I've just pushed a start at an alternative solution that I've been tinkering on for a few days, see #7881

**Comment by tmshort** [comment]
@levitte did your change make it in?

**Comment by tmshort** [comment]
Closing this in favor of @levitte's work.

**Pull request closed without merging**